packages/core/saved-objects/core-saved-objects-migration-server-internal/src/zdt/model/stages/init.ts

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import * as Either from 'fp-ts/lib/Either';
+import { delayRetryState } from '../../../model/retry_state';
+import { throwBadResponse } from '../../../model/helpers';
+import { isTypeof } from '../../actions';
+import type { State } from '../../state';
+import type { ModelStage } from '../types';
+
+export const init: ModelStage<'INIT'> = (state, res, context): State => {
```

Comment:
Allowing to define them just like that (parameter types are all properly inferred)

Reply:
Separating these into individual functions is really great! What do you think of something like this as an iteration on this idea:

```ts
// just an example
export const init: ModelTransition<'INIT', 'FATAL' | 'DONE'> = (state, res, context) => {}
```

In this way we will:

1. leverage TS sanity check on our expected return values
2. replace `State` return value declarations from these functions

`ModelTransition` would be defined as:

```ts
/**
 * Defines a transition function for the model
 */
export type ModelTransition<T extends AllActionStates, R extends AllControlStates> = (
  state: StateFromActionState<T>,
  res: StateActionResponse<T>,
  context: MigratorContext
) => StateFromControlState<T | R>; // T | R because a state can go back to itself...
```

Plus utility:

```ts
export type StateFromControlState<T extends AllControlStates> = ControlStateMap[T];
```

Reply:
I think your improvements are great. Done in https://github.com/elastic/kibana/pull/151282/commits/f7d6c7fe357030474dfb01ec55799526bfa56584